File: Chapter08/Code/test/selenium-tests.js
describe('Selenium tests', function () {
    var EC = protractor.ExpectedConditions;
    
    describe('homepage', function () {
        it('should cut off long titles', function () {
            browser.get('index.html');
            browser.wait(EC.presenceOf($('h3'), 1000));
            
            var elems = element.all(by.css('h3'));
            expect(elems.count()).toBe(3);
    
            var elem = elems.get(0);
            expect(elem.getCssValue('overflow')).toBe('hidden');
            expect(elem.getCssValue('text-overflow')).toBe('ellipsis');
            expect(elem.getCssValue('white-space')).toBe('nowrap');
        });
        
        it('should navigate to the clicked product', function () {
            browser.get('index.html');
            browser.wait(EC.presenceOf($('[ng-controller=homeController]'), 2000));
            
            var product = element.all(by.css('h3')).get(1);
            product.click();
            browser.wait(EC.presenceOf($('[ng-controller=productController]'), 2500));
            var title = $('h2');
            expect(title.getText()).toBe('Captain America: Civil War');
        });
        
        it('should search for all products containing "fanta"', function () {
            browser.get('index.html');
            browser.wait(EC.presenceOf($('[ng-controller=homeController]'), 2000));
            
            var input = element(by.css('[ng-model=query]'));
            input.sendKeys('fanta');            
            var searchBtn = element(by.css('#search-btn'));
            searchBtn.click();
            browser.wait(EC.presenceOf($('[ng-controller=searchController]'), 2500));
            var results = element.all(by.css('.thumbnail'));
            expect(results.count()).toBeLessThan(20);

            var allHaveFanta = results.reduce(function (acc, elem) {
                return elem.element(by.css('h3')).getText().then(function (caption) {
                    return acc && caption.toLowerCase().indexOf('fanta') > -1;
                });
            }, true);
            
            expect(allHaveFanta).toBeTruthy();
        });
        
        it('should remove an item from the shopping cart', function (done) {
            browser.get('shopping-cart.html');
            // redirect
            browser.wait(EC.presenceOf($('[ng-controller=loginController]'), 2000));
            
            var username = $('input[ng-model=username]');
            var password = $('input[type=password]');
            
            username.sendKeys('sander');
            password.sendKeys('1234');
            
            var submit = $('input[type=submit]');
            submit.click();
            // another redirect
            browser.wait(EC.presenceOf($('[ng-controller=shoppingCartController]'), 2000));
            
            browser.get('index.html');
            browser.wait(EC.presenceOf($('.btn.btn-primary'), 2000));
            
            var alert;
            var buttons = $$('.btn.btn-primary');
            
            buttons.get(0).click();
            browser.wait(EC.alertIsPresent(), 2000);
            alert = browser.switchTo().alert();
            alert.accept();
            
            buttons.get(1).click();
            browser.wait(EC.alertIsPresent(), 2000);
            var alert = browser.switchTo().alert();
            alert.accept();
            
            browser.get('shopping-cart.html');
            browser.wait(EC.presenceOf($('[ng-controller=shoppingCartController]'), 2000));
            
            browser.wait(EC.presenceOf($('.thumbnail'), 2000));
            
            var items = $$('.thumbnail');
            items.count().then(function (count) {
                var first = items.get(0);
                first.$('h3').getText().then(function (caption) {
                    first.$('button').click().then(function () {
                        // we don't really have anything to wait on...
                        // waiting for the element not to be there may cause
                        // an infinite loop, so just sleep for a second and continue.
                        browser.sleep(1000);
                        var newItems = $$('.thumbnail');
                        expect(newItems.count()).toBe(count - 1);
                        expect(newItems.get(0).$('h3').getText()).not.toBe(caption);
                        
                        // Refresh and check if the deleted item is still gone.
                        browser.refresh();
                        browser.wait(EC.presenceOf($('.thumbnail'), 2000));
            
                        var afterRefresh = $$('.thumbnail');
                        expect(afterRefresh.count()).toBe(count - 1);
                        done();
                    });
                });
            });
        });
    });
});

File: Chapter13/Code/web-shop-csharp/web-shop-selenium/SeleniumTests.cs
using Microsoft.AspNetCore.TestHost;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using System;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading;
using Xunit;
using web_shop;

namespace web_shop_selenium
{
    public class SeleniumTests : IDisposable
    {
        private readonly IWebDriver driver = new ChromeDriver();

        private readonly TestServer server;
        private readonly HttpClient client;

        public SeleniumTests()
        {
            server = new TestServer(new WebHostBuilder()
                .UseStartup<Startup>());
            client = server.CreateClient();
            client.BaseAddress = new Uri("http://ciserver:9999");
        }

        [Fact]
        public void ShouldCutOffLongTitles_Test()
        {
            driver.Navigate().GoToUrl($"{client.BaseAddress}");
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(2);
            var elems = driver.FindElements(By.CssSelector("h3"));
            Assert.Equal(3, elems.Count);

            var elem = elems[0];
            Assert.Equal("hidden", elem.GetCssValue("overflow"));
            Assert.Equal("ellipsis", elem.GetCssValue("text-overflow"));
            Assert.Equal("nowrap", elem.GetCssValue("white-space"));
        }

        [Fact]
        public void ShouldNavigateToTheClickedProduct_Test()
        {
            driver.Navigate().GoToUrl($"{client.BaseAddress}");
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(2);
            
            var product = driver.FindElements(By.CssSelector("h3"))[1];
            product.Click();

            var title = driver.FindElement(By.CssSelector("h2"));
            Assert.Equal("Captain America: Civil War", title.Text);
        }

        [Fact]
        public void ShouldSearchForAllProductsContainingFanta_Test()
        {
            driver.Navigate().GoToUrl($"{client.BaseAddress}");
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(2);
            
            var input = driver.FindElement(By.CssSelector("[ng-model=query]"));
            input.SendKeys("fanta");            
            var searchBtn = driver.FindElement(By.CssSelector("#search-btn"));
            searchBtn.Click();
            var results = driver.FindElements(By.CssSelector(".thumbnail"));
            Assert.True(results.Count < 20);

            var allHaveFanta = results.All(elem =>
            {
                var caption = elem.FindElement(By.CssSelector("h3")).Text;
                return caption.ToLowerInvariant().Contains("fanta");
            });
            
            Assert.True(allHaveFanta, "Not all found results contain the text 'fanta'.");
        }

        [Fact]
        public void ShouldRemoveAnItemFromTheShoppingCart_Test()
        {
            driver.Navigate().GoToUrl($"{client.BaseAddress}ShoppingCart");
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(2);
            
            // redirect            
            var username = driver.FindElement(By.CssSelector("input[ng-model=username]"));
            var password = driver.FindElement(By.CssSelector("input[type=password]"));
            
            username.SendKeys("user");
            password.SendKeys("password");
            
            var submit = driver.FindElement(By.CssSelector("input[type=submit]"));
            submit.Click();

            // another redirect
            driver.Navigate().GoToUrl($"{client.BaseAddress}");
            
            var buttons = driver.FindElements(By.CssSelector(".btn.btn-primary"));
            
            buttons[0].Click();
            driver.SwitchTo().Alert().Accept();
            
            buttons[1].Click();
            driver.SwitchTo().Alert().Accept();
            
            driver.Navigate().GoToUrl($"{client.BaseAddress}ShoppingCart");

            var items = driver.FindElements(By.CssSelector(".thumbnail"));
            var count = items.Count;
            var caption = items[0].FindElement(By.CssSelector("h3")).Text;
            items[0].FindElement(By.CssSelector("button")).Click();
            
            // Like with the JavaScript tests, just wait for a second...
            Thread.Sleep(1000);
            var newItems = driver.FindElements(By.CssSelector(".thumbnail"));
            Assert.Equal(count - 1, newItems.Count);
            Assert.NotEqual(caption, newItems[0].FindElement(By.CssSelector("h3")).Text);

            driver.Navigate().Refresh();
            var afterRefresh = driver.FindElements(By.CssSelector(".thumbnail"));
            Assert.Equal(count - 1, afterRefresh.Count);
        }

        #region IDisposable Support
        private bool disposedValue = false;

        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    driver.Dispose();
                    client.Dispose();
                    server.Dispose();
                }
                disposedValue = true;
            }
        }

        void IDisposable.Dispose()
        {
            Dispose(true);
        }
        #endregion
    }
}